### PR TITLE
IFU 2.12 PyTorch merge conflict resolution

### PR DESCRIPTION
**Pull request by Micky774**
# Description

This PR focuses on the resolutions to merge conflicts located in the PyTorch integration, as introduced as part of the main IFU branch. This merge resolution is performed in separate PRs to limit the scope of each PR and simplify the review process, while preserving commit history and conversation readability.

This is one of three PRs which target the common components of TE, as well as the JAX and PyTorch integrations. Once all three PRs are resolved and merged into the IFU branch, a new IFU PR will be created with a focus on ensuring a successful total build as well as CI validation. In the meanwhile, the PRs will be partially validated via local testing.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

The changelog is generated with assistance from Claude.

  The large negative delta (-590 lines) is because upstream refactored autograd forward() signatures to bundle arguments into a `non_tensor_args` tuple — removing the old per-parameter style that ROCm had conflict markers against.

  C++ / CSRC changes:

  1. `DeviceGuard` pattern — added `HIPGuardMasqueradingAsCUDA` conditional typedef to `attention.cpp`, `gemm.cpp`, and `normalization.cpp`, replacing `at::cuda::CUDAGuard` with the platform-aware `DeviceGuard`. This is necessary due to the inability to properly hipify the standard `at::cuda::CUDAGuard` to the proper masquerading HIP device guard.
  2. `cast.cpp` — guarded NVFP4-specific enum values (`BULK_NVFP4, FUSED_NVFP4`) and code paths with `#ifndef USE_ROCM`; moved the `#endif` for the bulk allocation function to the correct location
  3. `quantizer.cpp` — resolved scale shape computation conflicts: adopted upstream's `ceildiv()` helper while preserving ROCm's guard to skip `roundup(..., 4)` padding (ROCm doesn't need 4-aligned scale dimensions)
  4. `common.cpp / common.h` — kept both ROCm's atomic amax functions and upstream's new `ceildiv` utility; resolved `kFloat8E8M0` dtype mapping conflict keeping ROCm's FNUZ variant
  5. `extensions.h / util.h` — moved `#ifndef USE_ROCM` guard to correct positions, keeping `inplace_swizzle_scale_for_gemm` declaration visible while guarding NVSHMEM APIs

  Python module changes:

  6. `gemm.py` — added ROCm workspace sizing to `get_cublas_workspace_size_bytes()` (64 MiB for gfx950, 32 MiB for other ROCm GPUs) to avoid a HIPBLASLT error
  7. `grouped_linear.py`, `layernorm_linear.py`, `layernorm_mlp.py`, `linear.py` — adopted upstream's `non_tensor_args` tuple pattern for autograd forward, integrating ROCm-specific fields (`m_splits_tensor, use_grouped_gemm_triton, keep_fp8_weight_transpose_cache, use_fsdp2`) into the tuple. Removed the old per-parameter signatures and their corresponding `None` returns in backward
  8. `base.py` — kept the old `get_cublas_workspace_size_bytes()` (still used by legacy code paths) while integrating upstream's `NVFP4TensorStorage` import and `get_nvtx_range_context` utility
  9. `jit.py`, `distributed.py` — resolved trivial conflicts (import merges, trailing whitespace)
  10. `float8_blockwise_tensor.py` — resolved conflicts in scale shape computation, same pattern as `quantizer.cpp` (skip roundup padding on ROCm)
  11. Test changes — `test_attention.py` skips GQA with packed QKV layouts; `run_attention_with_cp.py` uses `IS_HIP_EXTENSION` to branch `cu_seqlens` handling; `test_numerics.py` keeps both ROCm wgrad skip and upstream's debug mode skip

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/cpp_extensions/gemm.py
Comment (ipanfilo) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:3]: AMD copyright

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 1]
```diff
@@ -29,8 +30,13 @@
 
 
 def get_cublas_workspace_size_bytes() -> None:
-    """Return 32 MiB if using hopper, 4 MiB for all other architectures."""
```
Comment (ipanfilo) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:32]: Return the comment to upstream code below

File: transformer_engine/pytorch/csrc/extensions/attention.cpp [hunk 2]
```diff
@@ -4,6 +4,14 @@
  * See LICENSE for license information.
  ************************************************************************/
 
+#ifdef USE_ROCM
+#include <ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h>
+using DeviceGuard = c10::hip::HIPGuardMasqueradingAsCUDA;
+#else
+#include <c10/cuda/CUDAGuard.h>
+using DeviceGuard = c10::cuda::CUDAGuard;
```
Comment (ipanfilo) [inline transformer_engine/pytorch/csrc/extensions/attention.cpp]: at::cuda
Also, it it appears in several places, rather add it to build_tools/hipify/custom_map.json

File: transformer_engine/pytorch/csrc/extensions/cast.cpp [hunk 3]
```diff
@@ -1154,8 +1154,20 @@ std::vector<py::object> split_quantize(const at::Tensor &tensor,
   }
 
   // Choose implementation for allocating and populating tensors
-  enum class AllocationMethod { UNFUSED, BULK_FP8_BLOCKWISE, BULK_MXFP8, BULK_NVFP4 };
-  enum class QuantizationMethod { UNFUSED, FUSED_NVFP4 };
+  enum class AllocationMethod {
+    UNFUSED,
+    BULK_FP8_BLOCKWISE,
+    BULK_MXFP8,
+#ifndef USE_ROCM
```
Comment (ipanfilo) [inline transformer_engine/pytorch/csrc/extensions/cast.cpp]: What is harm in having then unguarded?

File: transformer_engine/pytorch/module/base.py [hunk 4]
```diff
@@ -87,14 +84,13 @@ class UserBufferQuantizationMode(Enum):
     FP8 = "fp8"
 
 
-<<<<<<< HEAD
 def get_cublas_workspace_size_bytes() -> None:
```
Comment (ipanfilo) [inline transformer_engine/pytorch/module/base.py]: Aren't those methods removed from here by IFU

File: transformer_engine/pytorch/module/grouped_linear.py [hunk 5]
```diff
@@ -52,16 +52,13 @@
     prepare_for_saving,
     restore_from_saved,
 )
-<<<<<<< HEAD
 from torch.utils.cpp_extension import IS_HIP_EXTENSION
+from ...debug.pytorch.debug_quantization import DebugQuantizer
```
Comment (ipanfilo) [inline transformer_engine/pytorch/module/grouped_linear.py:55]: Nit: move them higher to have IS_HIP_EXTENSION import and use grouped together

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 6]
```diff
@@ -31,6 +31,7 @@
     _2X_ACC_WGRAD,
 )
 from ..quantization import FP8GlobalStateManager
+from ..tensor.float8_tensor import Float8CurrentScalingQuantizer
```
Comment (ipanfilo) [inline transformer_engine/pytorch/module/layernorm_linear.py]: Nit: it is better to be where it was, grouped with MXFP8Quantizer import

File: transformer_engine/pytorch/optimizers/multi_tensor_apply.py [hunk 7]
```diff
@@ -1,10 +1,6 @@
-<<<<<<< HEAD
 # This file was modified for portability to AMDGPU
 # Copyright (c) 2025-2026, Advanced Micro Devices, Inc. All rights reserved
 #  Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
-=======
-# Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (ipanfilo) [inline transformer_engine/pytorch/optimizers/multi_tensor_apply.py:6]: Nit: Leave this variant so it (probably) won't conflict with further upstream changes

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py [hunk 8]
```diff
@@ -166,16 +166,13 @@ def make_empty(
         if self.columnwise_usage:
             columnwise_data = torch.empty(
                 shape, dtype=torch.uint8, device=device, pin_memory=pin_memory
-<<<<<<< HEAD
-=======
             )
             columnwise_scale_inv = torch.empty(
```
Comment (ipanfilo) [inline transformer_engine/pytorch/tensor/mxfp8_tensor.py]: It duplicates code under else condition below

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/cpp_extensions/gemm.py
Comment (Micky774) [reply transformer_engine/pytorch/cpp_extensions/gemm.py:3]: Updated

File: transformer_engine/pytorch/cpp_extensions/gemm.py (hunk 1 above)
Comment (Micky774) [reply transformer_engine/pytorch/cpp_extensions/gemm.py:32]: Updated

File: transformer_engine/pytorch/csrc/extensions/attention.cpp (hunk 2 above)
Comment (Micky774) [reply transformer_engine/pytorch/csrc/extensions/attention.cpp]: Ah didn't know about the custom map, thanks! Updated

File: transformer_engine/pytorch/csrc/extensions/cast.cpp (hunk 3 above)
Comment (Micky774) [reply transformer_engine/pytorch/csrc/extensions/cast.cpp]: None -- removed the guard

File: transformer_engine/pytorch/module/base.py (hunk 4 above)
Comment (Micky774) [reply transformer_engine/pytorch/module/base.py]: Thought they were still used but it turns out to just be from our custom tests, so I've updated imports and removed them.

File: transformer_engine/pytorch/module/grouped_linear.py (hunk 5 above)
Comment (Micky774) [reply transformer_engine/pytorch/module/grouped_linear.py:55]: Updated

File: transformer_engine/pytorch/module/layernorm_linear.py (hunk 6 above)
Comment (Micky774) [reply transformer_engine/pytorch/module/layernorm_linear.py]: Updated

File: transformer_engine/pytorch/optimizers/multi_tensor_apply.py (hunk 7 above)
Comment (Micky774) [reply transformer_engine/pytorch/optimizers/multi_tensor_apply.py:6]: Updated

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py (hunk 8 above)
Comment (Micky774) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py]: Missed that, thanks. Updated

File: build_tools/hipify/custom_map.json [hunk 9]
```diff
@@ -8,7 +8,8 @@
         "<nvtx3/nvToolsExt.h>" : "<roctracer/roctx.h>",
         "cudaFuncSetAttribute(" : "hipFuncSetAttribute((const void*)",
         "__nv_bfloat162":"__hip_bfloat162",
-        "cuda::getCurrentCUDAStream" : "hip::getCurrentHIPStreamMasqueradingAsCUDA"
+        "cuda::getCurrentCUDAStream" : "hip::getCurrentHIPStreamMasqueradingAsCUDA",
+        "at::cuda::CUDAGuard" : "c10::hip::HIPGuardMasqueradingAsCUDA"
```
Comment (wangye805) [inline build_tools/hipify/custom_map.json:12]: Do we have at::hip::HIPGuard? Direct translating a aten object to a c10 object in pytorch looks weird

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 10]
```diff
@@ -8,9 +10,10 @@
 import os
 import functools
 import torch
+from torch.utils.cpp_extension import IS_HIP_EXTENSION
 import transformer_engine_torch as tex
 from ..constants import TE_DType
-from ..utils import get_sm_count, _empty_tensor
+from ..utils import get_sm_count, _empty_tensor, get_device_compute_capability
```
Comment (wangye805) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:16]: ROCm specific guard?

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 11]
```diff
@@ -29,8 +32,14 @@
...
 def get_cublas_workspace_size_bytes() -> None:
+    """Return workspace size needed for current architecture."""
+    if IS_HIP_EXTENSION:
+        """Return 64 MiB for gfx50x, 32 MiB for all other architectures."""
+        if get_device_compute_capability(0) == 95:
+            return 67_108_864
+        return 33_554_432
     """Return 32 MiB if using hopper, 4 MiB for all other architectures."""
-    if torch.cuda.get_device_properties(torch.cuda.current_device()).major >= 9:
+    if get_device_compute_capability(0) >= 90:
```
Comment (wangye805) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:42]: This branch is for NV upstream? Let's keep it as it is

File: transformer_engine/pytorch/csrc/extensions.h [hunk 12]
```diff
@@ -489,17 +489,14 @@ void fused_multi_row_padding(at::Tensor input, at::Tensor output,
...
                                std::vector<size_t> unpadded_input_row_list);
-<<<<<<< HEAD
-#ifndef USE_ROCM
-=======
 
 /***************************************************************************************************
  * Scale swizzling for GEMM
  **************************************************************************************************/
 
 void inplace_swizzle_scale_for_gemm(py::handle &tensor);
```
Comment (wangye805) [inline transformer_engine/pytorch/csrc/extensions.h:497]: Do we support this swizzle?

File: transformer_engine/pytorch/tensor/float8_blockwise_tensor.py [hunk 13]
```diff
@@ -159,53 +159,27 @@ def get_scale_shape(self, shape: Iterable[int], columnwise: bool) -> Tuple[int,
             scale_dim0 = (dim0 + self.block_len - 1) // self.block_len
             scale_dim1 = (dim1 + self.block_len - 1) // self.block_len
             if columnwise:
-<<<<<<< HEAD
-                outer = math.ceil(K / self.block_len)
                 if IS_HIP_EXTENSION:
```
Comment (wangye805) [inline transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:162]: I don't think we support fp8 blockwise tensor at all. We can revert our previous changes and keep the file unchanged as nv upstream

**Review by wangye805 — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/csrc/extensions.h (hunk 12 above)
Comment (Micky774) [reply transformer_engine/pytorch/csrc/extensions.h:497]: We don't use it anywhere, but it's exposed via `swizzle_scales_for_gemm_` through pybind

File: build_tools/hipify/custom_map.json (hunk 9 above)
Comment (Micky774) [reply build_tools/hipify/custom_map.json:12]: Updated

File: transformer_engine/pytorch/cpp_extensions/gemm.py (hunk 10 above)
Comment (Micky774) [reply transformer_engine/pytorch/cpp_extensions/gemm.py:16]: Updated

File: transformer_engine/pytorch/cpp_extensions/gemm.py (hunk 11 above)
Comment (Micky774) [reply transformer_engine/pytorch/cpp_extensions/gemm.py:42]: Updated

File: transformer_engine/pytorch/tensor/float8_blockwise_tensor.py (hunk 13 above)
Comment (Micky774) [reply transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:162]: Updated
